File: academic-system/src/main/java/org/example/academic/system/service/ReportService.java
package org.example.academic.system.service;

import org.example.academic.system.AcademicSystem;
import org.example.academic.system.model.AcademicClass;
import org.example.academic.system.model.Assessment;
import org.example.academic.system.repository.PersistenceConfiguration;

public class ReportService {

    private AcademicSystem academicSystem;

    public ReportService(AcademicSystem academicSystem) {
        this.academicSystem = academicSystem;
    }

    public void generateClassSummaryReport() {
        System.out.println("\n===== CLASS ASSESSMENT SUMMARY REPORT =====");
        if (academicSystem.getClasses().isEmpty()) {
            System.out.println("No classes registered.");
            return;
        }
        for (AcademicClass c : academicSystem.getClasses()) {
            System.out.println("\nClass: " + c.getCode() + " | " + c.getTitle());
            if (c.getAssessments().isEmpty()) {
                System.out.println("  No assessments registered.");
            } else {
                for (Assessment a : c.getAssessments()) {
                    System.out.println("  - " + a.getClass().getSimpleName()
                            + " | value=" + a.getValue()
                            + " | weight=" + a.getWeight());
                }
            }
        }
    }

    public void generateAssessmentWeightReport() {
        System.out.println("\n===== ASSESSMENT WEIGHT REPORT =====");
        if (academicSystem.getClasses().isEmpty()) {
            System.out.println("No classes registered.");
            return;
        }
        for (AcademicClass c : academicSystem.getClasses()) {
            double total = c.getAssessments().stream()
                    .mapToDouble(Assessment::getWeight).sum();
            String status = Math.abs(total - 1.0) < 0.001 ? "VALID ✓" : "INVALID ✗";
            System.out.println("Class: " + c.getCode() + " | " + c.getTitle()
                    + " | Total weight: " + total + " | " + status);
        }
    }

    public void generatePersistenceConfigurationReport() {
        System.out.println("\n===== PERSISTENCE CONFIGURATION REPORT =====");
        System.out.println("Current persistence type: "
                + PersistenceConfiguration.getCurrentType());
    }
}

File: academic-system-python/src/academic_system/service/report_service.py
from academic_system.academic_system import AcademicSystem
from academic_system.repository.persistence_configuration import PersistenceConfiguration
from academic_system.security.session import Session
import logging

logger = logging.getLogger(__name__)


class ReportService:
    def __init__(self, system: AcademicSystem):
        self._system = system

    def generate_class_summary_report(self) -> str:
        logger.info(f"Generating class summary report: "
                    f"role={Session.get_current_role()}")
        lines = ["===== CLASS ASSESSMENT SUMMARY REPORT ====="]
        classes = self._system.get_classes()
        if not classes:
            lines.append("No classes registered.")
            return "\n".join(lines)
        for c in classes:
            lines.append(f"\nClass: {c.code} | {c.title}")
            if not c.assessments:
                lines.append("  No assessments registered.")
            else:
                for a in c.assessments:
                    lines.append(f"  - {a.get_type()} | "
                                 f"value={a.value} | weight={a.weight}")
        return "\n".join(lines)

    def generate_assessment_weight_report(self) -> str:
        logger.info(f"Generating weight report: "
                    f"role={Session.get_current_role()}")
        lines = ["===== ASSESSMENT WEIGHT REPORT ====="]
        for c in self._system.get_classes():
            total = sum(a.weight for a in c.assessments)
            status = "VALID ✓" if abs(total - 1.0) < 0.001 else "INVALID ✗"
            lines.append(f"Class: {c.code} | {c.title} | "
                         f"Total weight: {total} | {status}")
        return "\n".join(lines)

    def generate_persistence_configuration_report(self) -> str:
        logger.info(f"Generating persistence config report: "
                    f"role={Session.get_current_role()}")
        return (f"===== PERSISTENCE CONFIGURATION REPORT =====\n"
                f"Current persistence type: "
                f"{PersistenceConfiguration.get_current_type().value}")